lib/rspec/matchers/built_in/compound.rb

```diff
@@ -36,6 +36,30 @@ def expects_call_stack_jump?
           NestedEvaluator.matcher_expects_call_stack_jump?(matcher_2)
         end
 
+        # @api private
+        # @return [Boolean]
+        def diffable?
+          matcher_1.diffable? || matcher_2.diffable?
+        end
+
+        # @api private
+        # @return [RSpec::Matchers::ExpectedsForMultipleDiffs]
+        def expected
+          return nil unless evaluator
+
+          expecteds = ::RSpec::Matchers::ExpectedsForMultipleDiffs.new
+
+          if !matcher_1_matches? && matcher_1.diffable?
+            expecteds.add(matcher_1.expected, :for_matcher => matcher_1)
+          end
+
+          if !matcher_2_matches? && matcher_2.diffable?
+            expecteds.add(matcher_2.expected, :for_matcher => matcher_2)
+          end
```

Comment:
Also, do you think it would make sense to pass the matcher description to `expecteds` rather than the matcher?  All it needs is the description, and passing a string feels like lower coupling to me.  If there's a concrete reason you think it's advantageous to pass the matcher itself I'm open to keeping it as is.